source/slang/ir-specialize.cpp

```diff
@@ -0,0 +1,684 @@
+    // we will try to be careful and only add things that aren't
+    // already fully specialized.
+    //
+    void maybeAddToWorkList(IRInst* inst)
+    {
+        if(isInstFullySpecialized(inst))
+            return;
+
+        workList.Add(inst);
+    }
+
+    // When we go to populate the work list by recursively
+    // traversing some code, we will be careful to *not*
+    // add generics or their children to the work list,
+    // and will instead consider a generic to be "fully
+    // specialized" already (because uses of that generic
+    // as an *operand* should be seen as fully specialized
+    // references).
+    //
+    void populateWorkListRec(
+        IRInst*                     inst)
+    {
+        if(auto genericInst = as<IRGeneric>(inst))
+        {
+            fullySpecializedInsts.Add(genericInst);
+        }
+        else
+        {
+            maybeAddToWorkList(inst);
+
+            for(auto child : inst->getChildren())
+            {
+                populateWorkListRec(child);
+            }
+        }
+    }
+
+    // Of course, somewhere along the way we expect
+    // to run into uses of `specialize(...)` instructions
+    // to bind a generic to arguments that we want to
+    // specialize into concrete code.
+    //
+    // We also know that if we encouter `specialize(g, a, b, c)`
+    // and then later `specialize(g, a, b, c)` again, we
+    // only want to generate the specialized code for `g<a,b,c>`
+    // *once*, and re-use it for both versions.
+    //
+    // We will cache existing specializations of generic function/types
+    // using the simple key type defined as part of the IR cloning infrastructure.
+    //
+    typedef IRSimpleSpecializationKey Key;
+    Dictionary<Key, IRInst*> genericSpecializations;
+
+    // We will also use some shared IR building state across
+    // all of our specialization/cloning steps.
+    //
+    SharedIRBuilder sharedBuilderStorage;
+
+    // Now let's look at the task of finding or generation a
+    // specialization of some generic `g`, given a specialization
+    // instruction like `specialize(g, a, b, c)`.
+    //
+    // The `specializeGeneric` function will return a value
+    // suitable for use as a replacement for the `specialize(...)`
+    // instruction.
+    //
+    IRInst* specializeGeneric(
+        IRGeneric*      genericVal,
+        IRSpecialize*   specializeInst)
+    {
+        // First, we want to see if an existing specialization
+        // has already been made. To do that we will construct a key
+        // for lookup in the generic specialization context.
+        //
+        // Our key will consist of the identity of the generic
+        // being specialized, and each of the argument values
+        // being pased to it. In our hypothetical example of
+        // `specialize(g, a, b, c)` the key will then be
+        // the array `[g, a, b, c]`.
+        //
+        Key key;
```

Comment:
Could setSize, and then just set the args to stop the array getting too big/being grown.

Reply:
True, although my instinct is to keep the code as simple as possible rather than pursue optimization opportunities like that which would obfuscate the code.

A better long-term option might be to use something like the LLVM project `SmallArray<T,N>` which includes direct storage for `N` elements so that you only need to do heap allocation for cases that go over some limit (at which point the slop of a stretchy dynamic array is maybe a minor worry). The 99% case for generics would probably have < 5 parameters, so even a small local array would avoid the whole problem.

Reply:
Having something like SmallArray would be beneficial in lots of places. The Key probably typically has a small amount of items in it too.